Add breadcrumbs to page title in admin backend

Murph33 opened the pull request:
Adding breadcrumbs will make for better UX.  As requested in #1076.  The button link to indexes have been taken out and replaced with navigatable breadcrumbs at the top of just about every page.

No new partials were added but that might be something we want to look at; pulling out the code in related pages when we can.  However in a lot of cases it would just be pulling 1-2 lines of code into a partial and then still requiring to add some page title code to that individual page.  

Just about every page was touched so the easiest way to get a feel for the changes might be to pull this down and just click around locally.

## Before Breadcrumbs

![no_breadcrumbs_1](https://cloud.githubusercontent.com/assets/12613649/14685849/32b4d488-06eb-11e6-9eff-6c2bf0dcb688.png)
![no_breadcrumbs_2](https://cloud.githubusercontent.com/assets/12613649/14685852/32b87cfa-06eb-11e6-9eb9-8680861fe9ed.png)
![no_breadcrumbs_3](https://cloud.githubusercontent.com/assets/12613649/14685850/32b7c4c2-06eb-11e6-85f4-e73c9bce9c5b.png)
![no_breadcrumbs_4](https://cloud.githubusercontent.com/assets/12613649/14685851/32b7e858-06eb-11e6-9fb8-51754b1b0314.png)

## After Breadcrumbs

![afterbread1](https://cloud.githubusercontent.com/assets/12613649/16249015/0497ff7c-37c8-11e6-8a37-6de0a74cc3e1.png)
![afterbread2](https://cloud.githubusercontent.com/assets/12613649/16249022/0d5c4cb2-37c8-11e6-8430-cfca3c92c362.png)
![afterbread3](https://cloud.githubusercontent.com/assets/12613649/16249024/0ed8174c-37c8-11e6-8b6a-62f52849ced9.png)
![afterbread4](https://cloud.githubusercontent.com/assets/12613649/16249025/0fb67f96-37c8-11e6-9650-21fee1ce696f.png)

tvdeyen commented on backend/app/views/spree/admin/orders/customer_details/edit.html.erb:
```diff
@@ -1,11 +1,10 @@
+  &#x2f; <%= Spree.t(:customer_details) %>
 <% end %>
 
 <% content_for :page_actions do %>
```
Shouldn't we remove all empty `content_for :page_actions` blocks? Or do we want to keep them for potential deface overrides?

Murph33 replied:
Something along those lines was my thought when I was going through but in hindsight I suppose since they're just empty `content_for`s with no text in them they wouldn't even affect the rendering of that element on the page or not.  I'm not sure if that will cause the `page_actions` block to still render on the page but I imagine we would want to keep it for the potential deface overrides.

tvdeyen replied:
Or just remove them. I don't think it's "good practice" to deface `content_for`s, right?

Murph33 replied:
Oh, my thought was that without some text in that `content_for` that then there would no longer be a `page_actions` element on the page and possibly people would be defacing that but I haven't checked this hypothesis.

tvdeyen replied:
Ah, you're absolutely right. We check with `content_for? :page_actions` and only render the div if something is present. We could remove the check or advice people to add `content_for :page_actions` to their overrides. I would vote for to go with latter. It was never guaranteed that this div exists anyway. Thoughts @jhawthorn ?

Murph33 replied:
Just checked this out.  Without adding a &nbsp; or something to the block the `<ul class='page-actions'>` element doesn't show up on the page.  I'm thinking we'd probably like to make sure that we don't lose those in case people are hooking into them.  I will make sure I change that.

tvdeyen commented:
Wow, it just took me quite some time to figure out where the breadcrumbs for order views came from. Maybe this is the opportunity to move it out of `order_tabs` partial into the views, no? Or at least rename the partial into something more understandable? Don't have a good name in my mind right now, though.

tvdeyen commented:
Besides that 👍 this is a great improvement!

Mandily commented:
I pulled this down to look at it and noticed one small thing. In the reports list the name is 'sales total' (singular) in the breadcrumbs it's plural - I don't care which it is, but can you make it consistent please?

Otherwise it looks good!

tvdeyen commented:
"She is called @Mandily sharp eyes"

Murph33 commented:
With `order_tabs` I'd rather rename that to something more descriptive then pull the information out to every page.  Partials being responsible for some of the breadcrumbs does make sense in some of the cases.

Murph33 commented:
I went with `Sales total` because the report list is calling an actual value on the `report` and changing that would seem out of scope but I think `Sales Totals` would make more sense in each place.

graygilmore commented on backend/app/views/spree/admin/adjustment_reasons/edit.html.erb:
```diff
@@ -1,11 +1,13 @@
 
 <% content_for :page_title do %>
-  <%= Spree.t(:editing_adjustment_reason) %>
+  <%= Spree.t(:settings) %> &#x2f;
```
Instead of writing `&#x2f;` all of the time I think it would be better if we used CSS for this. Something like:

```scss
.page-title > *:before {
  content: ">";
}
```

graygilmore commented:
I don't think I like having all of these things shoved into the `h1`. They should be separate things. Maybe the active item could be the `h1`? Either way I think the implementation needs to change a bit.

graygilmore commented on backend/app/helpers/spree/admin/navigation_helper.rb:
```diff
@@ -1,6 +1,39 @@
+      def render_breadcrumbs
+        add_page_title_to_breadcrumbs if content_for?(:page_title)
+        content_tag :ol, class: 'breadcrumb' do
+          safe_join breadcrumbs.collect { |level|
```
`do`/`end` for multi-line blocks

mamhoff replied:
Both of these are valueable comments, I'd like to see them addressed (`map` > `collect`, `do / end` for multi-line blocks).

jhawthorn replied:
`do`/`end` is actually a little more awkward here because `safe_join breadcrumbs.collect do |level|` will associate the block with `safe_join` rather than `collect`.

I agree with the hound, prefer `map` over `collect`.

Murph33 commented:
So @Sinetheta hacked something much nicer together on the hackdays and kindly kept me involved by giving me a PR which apparently github doesn't notify you of.  I finally noticed today when I asked him how it was coming along so it is merged now.  I see Gray already commenting so I won't ping him.  Hoping you're still giving the thumbs up @tvdeyen!

Murph33 commented:
I would just like to point out that the appearance of the breadcrumbs have changed since it's no longer in an H1 (which as Gray pointed out was a mistake in the first place) so if style opinionated people would chime in that would be appreciated.  All I did was bump the size some since they are in the old title's position and the small text size didn't look right to me.  I imagine this will get changed around some as we continue to make changes to the admin.

tvdeyen commented:
Is there a reason why we don't use `content_for` here? The API seams very unintuitive and tedious to me. 

```erb
<% content_for(:breadcrumbs) do %>
  <% link_to ... %>
  ...
<% end %>
```

Then when rendering the breadcrumbs we can take care of active state, etc.

Sinetheta commented:
@tvdeyen my thought was to get the breadcrumbs closer to meta-data. With a content_for we have no validation over what's being jammed up there. With an interface like `add_breadcrumb` I'm hoping that we can soon take this decision out of the views entirely and get the info from our navigation.

Currently we have a quasi configurable nav. Some of it is in [backend_configuration.rb](https://github.com/solidusio/solidus/blob/60ee1dffc7c3932cd7652cc2b4566f08a12f74b8/backend/app/models/spree/backend_configuration.rb) but mostly it's contained int he views. I've talked with @jhawthorn about how we might make the nav configurable and I think he liked this pattern https://github.com/codeplant/simple-navigation/wiki/Configuration

> Then when rendering the breadcrumbs we can take care of active state, etc.
How would we do this with a `content_for` approach?

tvdeyen commented:
I like the idea of having a configurable navigation. That would be a great integration point for extensions.

Murph33 commented:
This has been updated to use the plural_resource_name helper and rebased on master.  Yes/No @tvdeyen, @Mandily, @Sinetheta , @mamhoff, @jhawthorn, @graygilmore, et al?

Mandily commented:
I know you removed the H1 from the breadcrumbs to be more semantic, but could you add a font-size of 16px (or 1rem, whatever unit we use) to the breadcrumbs to give it a bit more prominence on the page? That will make it smaller than what it was to account for more horizontal space being used, but larger than the 13px it and the rest of the body text is now.

Murph33 commented:
Changed pictures in original post and bumped the font size.

tvdeyen commented on backend/app/views/spree/admin/cancellations/index.html.erb:
```diff
@@ -1,6 +1,7 @@
-<% content_for :page_title do %>
-  <%= link_to Spree.t(:order_number, number: @order.number), spree.edit_admin_order_path(@order) %> - <%= Spree.t(:cancel_inventory) %>
-<% end %>
+<% add_breadcrumb(plural_resource_name(Spree::Order), spree.admin_orders_path) %>
```
I think we miss a `link_to` here?

tvdeyen commented:
Besides lots of (useless) additional new lines 👍

Thinking about having a `breadcrumbs` helper that takes a block or array of entries, so we can spare the repetitive `add_breadcrumb` calls. We just call the `add_breadcrumb` for each of the entries then. 

Would be much more easy to use, no?

Otherwise, great addition. Like that very much from the UI perspective.

Murph33 commented:
I apologize while also blaming @Sinetheta for any unnecessary new lines that have appeared.  You were right @tvdeyen that did require a link_to in there, good catch.  I'm agreed with Kevin as well when you factor in the build up in partials we have that this while not necessarily being a perfect world implementation, is serviceable.   

Fixed the link_to and rebased on current master.

Murph33 commented:
Rebased on current master.  Would really appreciate if we could get this merged or decide on some other change as I know @Mandily is eager to have it in so we can continue moving forward.

@tvdeyen, @jhawthorn, thanks.

mamhoff commented:
I think this is good, but an intermediate step. Once we have a configurable navigation across levels, that should handle the breadcrumbs. It's better than what we had before though, so 👍 .

tvdeyen commented:
Yes, this is better as it's right now. 👍

jhawthorn commented:
There are several changes I think this needs but I will send them in a future PR.

Thank you @Murph33, @Sinetheta, and @Mandily
